util/a2e/atoe_utils.c

```diff
@@ -422,14 +436,8 @@ atoe_vsnprintf(char *str, size_t count, const char *fmt, va_list args)
 	if (fmt == NULL) {                                      /*ibm@029013*/
 		fmt = NULLCHARSTRING;                               /*ibm@029013*/
 	}                                                       /*ibm@029013*/
-	if (str == NULL) {
-		return ERROR_RETVAL;
-	}
-	str[0] = '\0';
-
 	this.buffer = str;
-	this.end = str + count - 1;
-	*this.end = '\0';          /* ensure null-termination in case of failure */
+	this.end = this.buffer + count;
```

Comment:
This should not assume that `count` is zero when `str` is `NULL`:
```c
	this.end = this.buffer + ((NULL != str) ? count : 0);
```

Reply:
We must assume that the provided `count` is less than or equal to the buffer capacity otherwise it would cause issues. I add a check to return -1 if the buffer is null and count is > 0 but I can't check if the provided not null buffer capacity is not less than count.

Reply:
In my testing with standard implementation on Mac, I get `EXC_BAD_ACCESS` error when buffer is null and count is not zero. Maybe we should let it fail like this instead of returning -1. what do you think @keithc-ca ?

Reply:
Looking around on the net, I couldn't find documentation for any implementation of `vsnprintf()` that permits the supplied buffer to be `NULL`. Given that, I'm not surprised that you encountered `EXC_BAD_ACCESS` (when count > 0).

On the one hand, that suggests all uses of `atoe_vsnprintf()` should be supplying a non-NULL buffer into which the function should write.

On the other hand, if we agree that this is a needed divergence from the API it replaces, then I suggest the behavior should follow the lead of `omrstr_vprintf()` which ignores the supplied buffer length when the target buffer is `NULL` (see [1]).

Can you share a link to documentation that explicitly allows the buffer to be `NULL`? If not, what is the motivation to have non-standard behavior for `atoe_vsnprintf()`?

[1] https://github.com/eclipse/omr/blob/master/port/common/omrstr.c#L997-L998

Reply:
as you mentioned, it is not mentioned in the manual that vsnprintf supports null buffer. However, the purpose of providing the `count` is to limit the number of chars to write. When count is zero, we can not write anything to the buffer therefore it should not care if the buffer is NULL or full or have space. it is the job of caller to make sure the buffer has `count` capacity and since null buffer has zero capacity it is totally acceptable to be provided.
Check this implementation for instance: https://github.com/torvalds/linux/blob/d2f51b3516dade79269ff45eae2a7668ae711b25/lib/vsprintf.c#L2749

Reply:
I can agree that it's reasonable to expect the caller to provide a reasonable `count`, the problem is that giving `count` priority over `buffer != NULL` yields behavior which differs from `omrstr_vprintf()` from this repository.

You didn't explain what motivates this change. What other changes should we anticipate that will depend on this behavior change?

Reply:
In fact the whole purpose of this PR is to add null support to the implementation because we use it with null buffer here: https://github.com/eclipse/omr/blob/2a9a7f4e479970496e279560224bdb7ce8d4c3a9/compiler/infra/String.cpp#L39 and it causes failure on zOS. the standard vsnprintf accepts null and fails with bad access exception if count is > 0.

Reply:
That motivation would have been a helpful addition to the description of this PR.

I suggest the title of this change should be adjusted: what it does is instead prioritize the buffer size over the buffer pointer (so if we don't need to actually write any characters, we don't care what that pointer is). What that, I'll review this again (probably tomorrow).

Reply:
Thanks @keithc-ca. I'll update the PR description as you suggested. please let me know if your requests are addressed correctly in my last commit. I'll merge commits after your check.